### PR TITLE
release:detach-tagged-base command initial implementation

### LINKED ISSUE #153: Implement "release:checkout-tagged-base" command
## Summary

`release:checkout-tagged-base` command should go through all the modules and check out the common base commit between `current branch` and `previous tagged` version.

## Reasoning

That command would allow us to operate already released versions adding some hand-picked patches on top of it, rather than working with everything committed into minor branches since the last release.

E.g. `4.5.0-RC1` contains the recipe we want to take and add a single critical patch, rather than using everything from the `4.5.x-dev` branches.

To achieve that we could use `git merge-base` for tag and branch.

### PR DESCRIPTION
**Pull request by dnsl48**
Resolves #153

### CODE REVIEW

File: src/Steps/Release/DetachTaggedBase.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+        foreach ($this->project->getAllChildren() as $lib) {
+            $repo = $lib->getRepository($output);
+
+            $path = $repo->getPath();
+
+            $planItem = $this->plan->getItem($lib->getName());
+
+            if (!$planItem) {
+                $this->log($output, "Skip the module {$lib->getName()} (couldn't find in the Plan)");
```
Comment (Cheddam) [inline src/Steps/Release/DetachTaggedBase.php]:
```suggestion
                $this->log($output, "Skipped module {$lib->getName()} (couldn't find in the Plan)");
```

File: src/Steps/Release/DetachTaggedBase.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                continue;
+            }
+
+            $libVersion = $planItem->getVersion();
+
+            if (!$libVersion) {
+                $this->log($output, "Skip the module {$lib->getName()} (couldn't determine the version)");
```
Comment (Cheddam) [inline src/Steps/Release/DetachTaggedBase.php]:
```suggestion
                $this->log($output, "Skipped module {$lib->getName()} (couldn't determine the version)");
```

**Comment by Cheddam** [comment]
Some notes from my testing against a `cwp/cwp-recipe-kitchen-sink`-based release:

- If a module is defined as a dependency of multiple recipes, this command will try to run the scan/checkout against it multiple times.
- The command tries to run against dependencies marked as 'upgrade-only' - this doesn't make sense, because we'd want to continue depending on the existing tag, rather than creating a new one.
- I'm not sure the logic is correct - for example, this plan had `silverstripe/auditor` set to `2.2.0`, and the command shifted the HEAD for that module to 2.1.5.

@dnsl48 Let's review this together when you're back.

**Comment by ScopeyNZ** [comment]
Any chance for some tests?

**Comment by Cheddam** [comment]
I had a crack at writing some test coverage for this operation, but struggled to produce a functioning set of mocks. With more time, I'd look into alternative approaches, but unfortunately I don't have the capacity at the moment.

Instead, I've raised a dedicated issue (#162) to look at producing test coverage for all of the `Steps`, or at the very least a base for more easily writing them in the future.

**Comment by dnsl48** [comment]
Yes, same story as in here - https://github.com/silverstripe/cow/pull/143#issuecomment-563445063

The current test scaffolding isn't in good shape at all, so writing tests become even more expensive than functionality.

**Pull request merged**